text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,648 @@
+>
+> Just to be clear, the Edge team is against implicit tail-calls for all functions, but they're in favor of tail-calls-with-an-explicit-keyword (similar to this RFC).
+
+
+An unofficial summary of the ECMA Script/ Javascript proposal for tail call/return
+https://github.com/carbon-language/carbon-lang/issues/1761#issuecomment-1198672079 (Jul, 2022)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC? -->
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+    - One point that needs to be decided is if TCE should be a feature that needs to be required from all backends or if it can be optional. Currently, the RFC specifies that an ICE should be issued if a backend cannot guarantee that TCE will be performed.
+    - Another point that needs to be decided is if TCE is supported by a backend what exactly should be guaranteed? While the guarantee that there is no stack growth should be necessary, should performance (as in transforming `call` instructions into `jmp`) also be guaranteed? Note that a backend that guarantees performance should do so **always** otherwise the main intent of this RFC seems to be lost.
+    - Migration guidance, it might be interesting to provide a lint that indicates that a trivial transformation from `return` to `become` can be done for function calls where all requisites are already fulfilled. However, this lint might be confusing and noisy. Decide on if this lint or others should be added.
+    - Should a lint be added for functions that are marked to be a tail call or use become. See the discussion [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1159822824), as well as, the clippy and rustfmt changes of an initial [implementation](https://github.com/semtexzv/rust/commit/29f430976542011d53e149650f8e6c7221545207#diff-6c8f5168858fed7066e1b6c8badaca8b4a033d0204007b3e3025bf7dd33fffcb) (2022).
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+    - Are all calling-convention used by Rust available for TCE with the proposed restrictions on function signatures?
+    - Can the restrictions on function signatures be relaxed?
+        - One option for intra-crate direct calls is to automatically pad the arguments during compilation see [here](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1500620309). Does this have an influence on other calls? How much implementation effort is it?
+    - Can generic functions be supported?
```

Comment:
Given that Rust uses monomorphization, I don't see any reason why generic functions may not be supported.

Reply:
You are right, generic functions should not be a problem. Just for context the list here is of all features that I have found to be discussed in the context of tail calls. So if the answer to some of them is that there is no problem, then this was just me being overly cautious.

Reply:
Thanks for being cautious! I think, however, that we should resolve the things we can, to make the list shorter and easier to grasp/focus on real questions.

Reply:
Agreed, just didn't get to it yet. Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/eeaa80c6db06ab1e9a7ecb755844e8b3da7baef8